File: src/main/java/com/nicta/scoobij/impl/DListImpl.java
/**
 * Copyright 2011 National ICT Australia Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.nicta.scoobij.impl;

import java.util.Vector;

import com.nicta.scoobij.DGroupedTable;
import com.nicta.scoobij.DList;
import com.nicta.scoobij.DTable;
import com.nicta.scoobij.DoFn;
import com.nicta.scoobij.Filterer;
import com.nicta.scoobij.FlatMapper;
import com.nicta.scoobij.Mapper;
import com.nicta.scoobij.OrderedWireFormatType;
import com.nicta.scoobij.TableFlatMapper;
import com.nicta.scoobij.TableMapper;
import com.nicta.scoobij.WireFormatType;
import com.nicta.scoobij.WireFormats;

public class DListImpl<A> implements com.nicta.scoobij.DList<A> {

	public DListImpl(com.nicta.scoobi.DList<A> dl) {
		impl = dl;
	}

	@Override
	public <B> DList<B> parallelDo(DoFn<A, B> fun, WireFormatType<B> wf) {
		return new DListImpl<B>(getImpl().parallelDo(Conversions.toScala(fun),
				wf.typeInfo(), wf.wireFormat()));
	}

	@Override
	public <V> DListImpl<V> flatMap(FlatMapper<A, V> fm,
			WireFormatType<V> bundle) {
		return new DListImpl<V>(impl.flatMap(Conversions.toScala(fm),
				bundle.typeInfo(), bundle.wireFormat()));
	}

	@Override
	public <B, C> DTable<B, C> tableFlatMap(TableFlatMapper<A, B, C> tm,
			WireFormatType<B> bundleB, WireFormatType<C> bundleC) {

		scala.reflect.Manifest<scala.Tuple2<B, C>> manifest = Conversions
				.toManifest(scala.Tuple2.class,
						bundleB.typeInfo(),
						bundleC.typeInfo());

		return new DTableImpl<B, C>(impl.flatMap(
				Conversions.toScala(tm),
				manifest,
				WireFormats.wireFormatPair(bundleB, bundleC).wireFormat()));
	}

	// ++ from scoobi.DList
	@Override
	public DListImpl<A> concat(DList<A>... dLists) {

		return new DListImpl<A>(impl.$plus$plus(toImplArray(dLists)));
	}

	private static <T> scala.collection.Seq<com.nicta.scoobi.DList<T>> toImplArray(
			DList<T>[] v) {

		Vector<com.nicta.scoobi.DList<T>> vecta = new Vector<com.nicta.scoobi.DList<T>>(
				v.length);

		for (int i = 0; i < v.length; ++i) {
			vecta.add(v[i].getImpl());
		}

		return scala.collection.JavaConversions.asScalaBuffer(vecta);
	}

	@Override
	public <B> DList<B> map(Mapper<A, B> mapper, WireFormatType<B> wf) {

		return new DListImpl<B>(getImpl().parallelDo(
				Conversions.toScalaDoFn(mapper),
				wf.typeInfo(), wf.wireFormat()));
	}

	@Override
	public <B, C> DTable<B, C> tableMap(TableMapper<A, B, C> tm,
			WireFormatType<B> bundleB, WireFormatType<C> bundleC) {
		return tableFlatMap(Conversions.toTableFlatMapper(tm), bundleB, bundleC);
	}

	@Override
	public DList<A> filter(Filterer<A> filter, WireFormatType<A> bundle) {
		return flatMap(Conversions.toFlatMapper(filter, false), bundle);
	}

	@Override
	public DList<A> filterNot(Filterer<A> filter, WireFormatType<A> bundle) {
		return flatMap(Conversions.toFlatMapper(filter, true), bundle);
	}

	@Override
	public <K> DGroupedTable<K, A> groupBy(Mapper<A, K> mapper,
			OrderedWireFormatType<K> bundleK, WireFormatType<A> bundleA) {
		return tableMap(Conversions.toByMap(mapper), bundleK, bundleA)
				.groupByKey(bundleK, bundleA);
	}

	@Override
	public <K> DTable<K, A> by(Mapper<A, K> mapper, WireFormatType<K> bundleK,
			WireFormatType<A> bundleA) {
		return tableMap(Conversions.toByMap(mapper), bundleK, bundleA);
	}

	public com.nicta.scoobi.DList<A> getImpl() {
		return impl;
	}

	private com.nicta.scoobi.DList<A> impl;

}

File: src/main/scala/com/nicta/scoobi/impl/plan/DListImpl.scala
/**
 * Copyright 2011,2012 National ICT Australia Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.nicta.scoobi
package impl
package plan

import org.apache.hadoop.io.compress.CompressionCodec
import org.apache.hadoop.io.SequenceFile.CompressionType
import com.nicta.scoobi.impl.plan.comp._
import com.nicta.scoobi.impl.plan.comp.Load
import CollectFunctions._

import core._
import comp._
import mapreducer._
import collection.Seqs._
import WireFormat._
import WireFormat._
import com.nicta.scoobi.core
import com.nicta.scoobi.impl.control.Exceptions._
import com.nicta.scoobi.impl.plan.comp.Load
import com.nicta.scoobi.impl.plan.comp.Materialise
import com.nicta.scoobi.impl.plan.comp.Combine
import com.nicta.scoobi.impl.plan.comp.GroupByKey

/** An implementation of the DList trait that builds up a DAG of computation nodes. */
private[scoobi]
class DListImpl[A](comp: ProcessNode) extends DList[A] {

  type C = ProcessNode

  def getComp: C = comp

  import scalaz.Store

  def storeComp: Store[C, DList[A]] =
    Store(new DListImpl[A](_), comp)

  def addSink(sink: Sink) =
    storeComp puts (_.addSink(sink))

  def updateSinks(f: Seq[Sink] => Seq[Sink]) =
    storeComp puts (_.updateSinks(f))

  def compressWith(codec: CompressionCodec, compressionType: CompressionType = CompressionType.BLOCK) =
    storeComp puts ((c: C) => c.updateSinks(sinks => sinks.updateLast(_.compressWith(codec, compressionType))))

  def parallelDo[B : WireFormat, E : WireFormat](env: DObject[E], dofn: EnvDoFn[A, B, E]): DList[B] =
    new DListImpl(ParallelDo.create(Seq(comp), env.getComp, dofn, wireFormat[A], wireFormat[B]))

  def ++(ins: DList[A]*): DList[A] = DListImpl.apply {
    val others = ins.map(_.getComp)
    comp match {
      // special case. If we just append processing nodes linked to sources
      // then we can create only one parallelDo node
      // this will avoid stack overflows when optimising the graph
      case ParallelDo(nodes, env, dofn, wa, wb, sinks, _) if others.forall(isParallelDo) =>
        val pds = others.collect(isAParallelDo)
        if (Return.isUnit(env)        && pds.forall(p => Return.isUnit(p.env)) &&
            comp.nodeSinks.isEmpty    && pds.forall(_.nodeSinks.isEmpty) &&
            dofn == EmitterDoFunction && pds.forall(_.dofn == EmitterDoFunction))
          nodes.toVector ++ pds.toVector
        else getComp +: others.toVector

      case other => other +: others.toVector
    }
  }

  def groupByKey[K, V]
      (implicit ev:   A <:< (K, V),
                wfk: WireFormat[K],
                gpk:  Grouping[K],
                wfv: WireFormat[V]): DList[(K, Iterable[V])] =
    new DListImpl(GroupByKey(comp, wfk, gpk, wfv))

  def combine[K, V]
      (f: Reduction[V])
      (implicit ev:   A <:< (K,Iterable[V]),
                wfk: WireFormat[K],
                wfv: WireFormat[V]): DList[(K, V)] =
    new DListImpl(Combine(comp, Combine.reducer((a1: Any, a2: Any) => f(a1.asInstanceOf[V], a2.asInstanceOf[V])), wfk, wfv, wfv))

  /**
   * Low-level combine to be able to emit arbitrary reduced values when the combiner is called
   */
  def combineDo[K, V, U](dofn: DoFn[Iterable[V], U])(implicit ev: A <:< (K, Iterable[V]), wfk: WireFormat[K], wfv: WireFormat[V], wfu: WireFormat[U]): DList[(K, U)] =
    new DListImpl(Combine(comp, dofn, wfk, wfv, wfu))

  lazy val materialise: DObject[Iterable[A]] = new DObjectImpl(Materialise(comp, wireFormat[Iterable[A]]))

  def parallelDo[B : WireFormat](dofn: DoFn[A, B]): DList[B] = parallelDo(UnitDObject.newInstance, dofn)

  override def toString = "\n"+ new ShowNode {}.pretty(comp)
}

private[scoobi]
object DListImpl {
  def apply[A](source: DataSource[_,_, A])(implicit wf: WireFormat[A]): DListImpl[A] = apply(Vector(Load(source, wf)))
  def apply[A](ins: Seq[CompNode])(implicit wf: WireFormat[A]): DListImpl[A] =  new DListImpl[A](ParallelDo.create(ins:_*)(wf))
}
